src/misc/umap.c

```diff
@@ -0,0 +1,864 @@
+
+    return IGRAPH_SUCCESS;
+}
+
+/*xd is difference in x direction, mu is a weight */
+/* NOTE: mu is the probability of a true edge in high dimensions, not affected
+ * by the embedding (in particular, xd and yd), so it's a constant for the
+ * derivative/force. Same applies for the repulsion */
+static igraph_error_t igraph_i_umap_attract(igraph_real_t xd, igraph_real_t yd, igraph_real_t mu,
+       igraph_real_t a, igraph_real_t b, igraph_real_t *force_x, igraph_real_t *force_y)
+{
+    igraph_real_t dsq, phi, force;
+
+    dsq = xd * xd + yd * yd;
+
+    phi = 1. / (1. + a * pow(dsq, b));
+    force = - mu * (2 * a * b * pow(dsq, b - 1)) * phi;
+    *force_x = force * xd;
+    *force_y = force * yd;
+
+    /* clip force to avoid too rapid change */
+    igraph_i_umap_clip_force(force_x, 3);
+    igraph_i_umap_clip_force(force_y, 3);
+
+#ifdef UMAP_DEBUG
+    printf("force attractive: xd = %f, fx = %f\n", xd, *force_x);
+    printf("force attractive: yd = %f, fy = %f\n", yd, *force_y);
+#endif
+
+    return IGRAPH_SUCCESS;
+}
+
+/*xd is difference in x direction, mu is a weight */
+static igraph_error_t igraph_i_umap_repel(igraph_real_t xd, igraph_real_t yd, igraph_real_t mu,
+       igraph_real_t a, igraph_real_t b, igraph_real_t *force_x, igraph_real_t *force_y)
+{
+    igraph_real_t dsq, force;
+    igraph_real_t min_dist = 0.01;
+
+    dsq = fmax(min_dist * min_dist, xd * xd + yd * yd);
+
+    /* NOTE: in practice, in negative sampling mu is always zero because we
+     * *assume* the sample to be negative i.e. never a true edge */
+    force = (1 - mu) * (2 * b) / dsq / (1 + a * pow(dsq, b));
+    *force_x = force * xd;
+    *force_y = force * yd;
+
+    /* clip force to avoid too rapid change */
+    igraph_i_umap_clip_force(force_x, 3);
+    igraph_i_umap_clip_force(force_y, 3);
+
+#ifdef UMAP_DEBUG
+    printf("force repulsive: xd = %f, fx = %f\n", xd, *force_x);
+    printf("force repulsive: yd = %f, fy = %f\n", yd, *force_y);
+#endif
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t igraph_i_umap_apply_forces(const igraph_t *graph,  const igraph_vector_t *umap_weights,
+       igraph_matrix_t *layout, igraph_real_t a, igraph_real_t b, igraph_real_t prob,
+       igraph_real_t learning_rate, igraph_bool_t avoid_neighbor_repulsion)
+{
+    igraph_integer_t no_of_nodes = igraph_matrix_nrow(layout);
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    igraph_integer_t from, to;
+    igraph_real_t from_x, from_y, to_x, to_y, x_diff, y_diff, fx, fy;
+    /* The following is only used for small graphs, to avoid repelling your neighbors
+     * For large sparse graphs, it's not necessary. For large dense graphs, you should
+     * not be doing UMAP.
+     * */
+    igraph_vector_int_t neis;
+
+    if (avoid_neighbor_repulsion) {
+        IGRAPH_VECTOR_INT_INIT_FINALLY(&neis, 0);
+    }
+
+    igraph_integer_t n_random_vertices = sqrt(no_of_nodes);
+
+    /* iterate over a random subsample of edges */
```

Comment:
We have a function for efficiently selecting a random subset, `igraph_random_sample()`.

Reply:
implemented, thank you